src/Renci.SshNet/Common/Pipe.cs

```diff
@@ -0,0 +1,47 @@
+﻿namespace Renci.SshNet.Common
+{
+    using System;
+    using System.IO;
+
+    /// <summary>
+    /// A generic pipe to pass through data.
+    /// </summary>
+    internal class Pipe : IDisposable
+    {
+        private readonly LinkedListQueue<byte[]> _queue;
+
+        /// <summary>
+        /// Gets the input stream.
+        /// </summary>
+        /// <value>The input stream.</value>
+        public Stream InputStream { get; private set; }
+
+        /// <summary>
+        /// Gets the output stream.
+        /// </summary>
+        /// <value>The output stream.</value>
+        public Stream OutputStream { get; private set; }
+
+        public Pipe()
+        {
+            _queue = new LinkedListQueue<byte[]>();
+            InputStream = new PipeInputStream(_queue);
```

Comment:
The questions I had were:

- should we immediately create the PipeInputStream, or postpone this until a client is actually requesting one (eg. postpone creating the instance until a client invokes SshCommand.OutputStream)?
- when we postpone creating the PipeInputStream until a client requests it (client could be SSH.NET itself), then shouldn't that client be responsible for disposing this stream (eg. the stream returned by SshCommand.OutputStream)?
- should we have a way to signal write errors to the reader? eg. do we just let PipeInputStream.Read() return 0 when the SSH session has been closed unexpectedly? one of such errors could also be that the writer was disposed before it signaled EOS.

To me, multiple disposes should never be an error. However In this case, that's not what I was trying to say.
See questions above.